sips/sip-018/sip-018-signed-structured-data.md

```diff
@@ -0,0 +1,210 @@
+
+(define-read-only (verify-signature (hash (buff 32)) (signature (buff 65)) (signer principal))
+	(is-eq (principal-of? (unwrap! (secp256k1-recover? hash signature) false)) (ok signer))
+)
+
+(define-read-only (verify-signed-structured-data (structured-data-hash (buff 32)) (signature (buff 65)) (signer principal))
+	(verify-signature (sha256 (concat structured-data-header structured-data-hash)) signature signer)
+)
+```
+
+# Related work
+
+- [EIP712](https://eips.ethereum.org/EIPS/eip-712) is a standard in Ethereum (and adopted by other EVM-compatible chains) describing a similar mechanism that enables signing of structured data. Some aspects of this SIP were loosely inspired by the standard.
+- [EIP191](https://eips.ethereum.org/EIPS/eip-191) now describes various signing schemes used in the EVM space. When `personal_sign` was introduced, it allowed users to sign data without running the risk of inadvertently signing a transaction by adding a prefix to the message. The authors were inspired by Bitcoin and thus settled on a message structure of `<varint_prefix_length><prefix><varint_message_length><message>`. For Ethereum, it effectively took the form of `\x19Ethereum Signed Message:\n`, followed by the message length, followed by the message. The `\x19` byte correctly encodes the prefix length, which is `25`. However,  the message length was [mistakenly](https://github.com/ethereum/go-ethereum/issues/14794) encoded as an ASCII string and not an actual `varint`. It makes verifying messages signed with `personal_sign` needlessly expensive on-chain, as the message byte length has to be converted to an ASCII string representation before the message hash can be obtained. What is more, the messages signed with `personal_sign` do not have a predefined structure which makes it very difficult for wallet applications to display them properly. [MetaMask](https://docs.metamask.io/guide/signing-data.html#a-brief-history), a popular EVM browser wallet extension, assesses if a message passed to `personal_sign` is valid UTF-8, and will display it as human-readable text if so. Otherwise, it converts the message to a hex string and displays that instead. Complex messages cannot be meaningfully expressed in UTF-8 whilst still allowing smart contracts to process them efficiently. Applications developers thus often resorted to requesting users to sign hashes of complex messages, making the data being signed completely opaque to the user. Considering these problems and Clarity's inherent readability, there is no reason for there to ever be a `personal_sign` equivalent in Stacks.
+
+# Backwards Compatibility
+
+Not applicable
+
+# Activation
+
+This SIP will be considered activated if the following two conditions are met:
+
+1. `to-consensus-buff` or equivalent functionality is added to Stacks 2.1.
+2. At least one popular wallet application implements the standard before or within three months of the launch of Stacks 2.1.
```

Comment:
You can easily give yourself a year for this if you wanted.

Reply:
Fair. And I guess it might become a moot point since Hiro has since begun implementation. (Although it is not yet according to the standard.)
- https://github.com/hirosystems/connect/issues/225